cluster_pack/filesystem.py

```diff
@@ -88,6 +91,62 @@ def ensure_bytes(self, s):
     def write(self, data):
         self.base_hdfs_file.write(self.ensure_bytes(data))
 
+    def _seek_delimiter(file, delimiter, blocksize=2 ** 16):
+        """ Seek current file to next byte after a delimiter
+        from https://github.com/dask/hdfs3/blob/master/hdfs3/utils.py#L11
+
+        Parameters
+        ----------
+        file: a file
+        delimiter: bytes
+            a delimiter like b'\n'
+        blocksize: int
+            number of bytes to read
+        """
+        last = b''
+        while True:
+            current = file.read(blocksize)
+            if not current:
+                return
+            full = last + current
+            try:
+                i = full.index(delimiter)
+                file.seek(file.tell() - (len(full) - i) + len(delimiter))
+                return
+            except ValueError:
+                pass
+            last = full[-len(delimiter):]
+
+    def readline(self, size=None):
+        """ Read and return a line of bytes from the file.
+
+        Line terminator is always b"\\n".
+
+        Parameters
+        -----------
+
+        size: int maximum number of bytes read until we stop
+
+        """
+        start = self.tell()
+        self._seek_delimiter(self.ensure_bytes("\n"))
+        end = self.tell()
+        self.seek(start)
+        return self.read(end - start)
+
+    def readlines(self, hint=None):

```

Comment:
Why not take the signature of the regular python readlines (no hint). It seems better plus what is the behavior when hint is a size which ends at the middle of one line?

Reply:
Turns out those are the Python FS functions:
```
 |  readline(self, size=-1, /)
 |      Read until newline or EOF.
 |
 |      Returns an empty string if EOF is hit immediately.
 |  readlines(self, hint=-1, /)
 |      Return a list of lines from the stream.
 |
 |      hint can be specified to control the number of lines read: no more
```

I added unit tests for different values of hint.

Reply:
https://docs.python.org/3.3/library/io.html#io.IOBase.readlines